pkg/kademlia/kademlia.go

```diff
@@ -0,0 +1,508 @@
+				k.waitNextMu.Unlock()
+
+				k.depthMu.Lock()
+				k.depth = k.recalcDepth()
+				k.depthMu.Unlock()
+
+				k.logger.Debugf("connected to peer: %s old depth: %d new depth: %d", peer, currentDepth, k.NeighborhoodDepth())
+
+				select {
+				case <-k.quit:
+					// shutting down
+					return true, false, nil
+				default:
+				}
+
+				// the bin could be saturated or not, so a decision cannot
+				// be made before checking the next peer, so we iterate to next
+				return false, false, nil
+			})
+
+			if err != nil {
+				if errors.Is(err, errMissingAddressBookEntry) {
+					po := uint8(swarm.Proximity(k.base.Bytes(), peerToRemove.Bytes()))
+					k.knownPeers.Remove(peerToRemove, po)
+				} else {
+					k.logger.Errorf("kademlia manage loop iterator: %v", err)
+				}
+			}
+		}
+	}
+}
+
+// binSaturated indicates whether a certain bin is saturated or not.
+// when a bin is not saturated it means we would like to proactively
+// initiate connections to other peers in the bin.
+func binSaturated(bin, depth uint8, peers *pslice.PSlice) bool {
+	// short circuit for bins which are >= depth
+	if bin >= depth {
+		return false
+	}
+
+	// lets assume for now that the minimum number of peers in a bin
+	// would be 2, under which we would always want to connect to new peers
+	// obviously this should be replaced with a better optimization
+	// the iterator is used here since when we check if a bin is saturated,
+	// the plain number of size of bin might not suffice (for example for squared
+	// gaps measurement)
+
+	size := 0
+	_ = peers.EachBin(func(_ swarm.Address, po uint8) (bool, bool, error) {
+		if po == bin {
+			size++
+		}
+		return false, false, nil
+	})
+
+	return size >= 2
+}
+
+// recalcDepth returns the kademlia depth. Must be called under lock.
+func (k *Kad) recalcDepth() uint8 {
+	// handle edge case separately
+	if k.connectedPeers.Length() <= nnLowWatermark {
+		return 0
+	}
+	var (
+		peers                        = uint(0)
+		candidate                    = uint8(0)
+		shallowestEmpty, noEmptyBins = k.connectedPeers.ShallowestEmpty()
+	)
+
+	_ = k.connectedPeers.EachBin(func(_ swarm.Address, po uint8) (bool, bool, error) {
+		peers++
+		if peers >= nnLowWatermark {
+			candidate = po
+			return true, false, nil
+		}
+		return false, false, nil
+	})
+
+	if noEmptyBins || shallowestEmpty > candidate {
+		return candidate
+	}
+
+	return shallowestEmpty
+}
+
+// connect connects to a peer and gossips its address to our connected peers,
+// as well as sends the peers we are connected to to the newly connected peer
+func (k *Kad) connect(ctx context.Context, peer swarm.Address, ma ma.Multiaddr, po uint8) error {
+	_, err := k.p2p.Connect(ctx, ma)
+	if err != nil {
+		k.logger.Debugf("error connecting to peer %s: %v", peer, err)
+		k.waitNextMu.Lock()
+		k.waitNext[peer.String()] = time.Now().Add(timeToRetry)
+		k.waitNextMu.Unlock()
+
+		// TODO: somehow keep track of attempts and at some point forget about the peer
```

Comment:
Just a comment. I think we should probably keep some info about the peers in a peer struct, and maybe even persisted in the addressbook as well. This way we can keep track of such things (blacklisting, trust, etc...). Just my 2 c.

Reply:
yes, but this is not trivial and i wanted to avoid implementing this on the first phase (and asking people to review more components with more complexity by doing so)

Reply:
I completely agree, just commented cuz I saw the comment!